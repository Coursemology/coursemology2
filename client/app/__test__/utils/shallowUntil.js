import { shallow } from 'enzyme';

// See https://github.com/airbnb/enzyme/issues/539 and the `until` helper was borrowed from there.
function until(selector, options) {
  let context = options && options.context;
  if (
    !selector ||
    this.isEmptyRender() ||
    typeof this.getElement().type === 'string'
  ) {
    return this;
  }

  const instance = this.getElement();
  if (instance.getChildContext) {
    context = {
      ...context,
      ...instance.getChildContext(),
    };
  }

  return this.is(selector)
    ? this.shallow({ context })
    : until.call(this.shallow({ context }), selector, { context });
}

/**
 * Shallow renders the component until the component matches the selector.
 * This is useful when the component you want to test is nested inside another component.
 * example:
 * ```
 * const component = <Provider>
 *                     <MyComponent />
 *                   </Provider>
 * ```
 * In the above case, `shallow(component)` will render the <Provider />, and
 * `shallowUntil(component, 'MyComponent')` will render <MyComponent />
 */
export default function shallowUntil(component, options, selector) {
  if (selector === undefined) {
    // eslint-disable-next-line no-param-reassign
    selector = options;
    // eslint-disable-next-line no-param-reassign
    options = undefined;
  }
  return until.call(shallow(component, options), selector, options);
}
