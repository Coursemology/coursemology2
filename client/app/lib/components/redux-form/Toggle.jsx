import { forwardRef } from 'react';
import PropTypes from 'prop-types';
import {
  FormControl,
  FormControlLabel,
  FormHelperText,
  Switch,
} from '@material-ui/core';
import createComponent from './createComponent';
import mapError from './mapError';

const propTypes = {
  checked: PropTypes.bool,
  disabled: PropTypes.bool,
  errorText: PropTypes.node,
  label: PropTypes.node,
  onToggle: PropTypes.func,
};

const styles = {
  toggleFieldStyle: {
    height: '30px',
    margin: '8px 0px 0px -8px',
  },
  errorText: { margin: 0 },
};

const renderToggleField = forwardRef((props, ref) => {
  const { checked, disabled, errorText, label, onToggle, ...custom } = props;
  const isError = !!errorText;
  return (
    <FormControl
      disabled={disabled}
      error={isError}
      fullWidth
      variant="standard"
    >
      <FormControlLabel
        control={
          <Switch checked={checked} color="primary" onChange={onToggle} />
        }
        label={<b>{label}</b>}
        {...custom}
        ref={ref}
        style={styles.toggleFieldStyle}
      />
      {isError && (
        <FormHelperText error={isError} style={styles.errorText}>
          {errorText}
        </FormHelperText>
      )}
    </FormControl>
  );
});

renderToggleField.displayName = `ToggleField`;
renderToggleField.name = 'ToggleField';
renderToggleField.propTypes = propTypes;

const mapProps = ({ input: { value, onChange, ...inputProps }, ...props }) => ({
  ...mapError({ ...props, input: inputProps }),
  checked: !!value,
  onToggle: onChange,
});

export default createComponent(renderToggleField, mapProps);
