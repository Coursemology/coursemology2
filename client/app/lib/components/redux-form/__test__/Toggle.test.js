import shallowUntil from 'testUtils/shallowUntil';
import Toggle from '../Toggle';

describe('<Toggle />', () => {
  it('renders a toggled toggle', () => {
    const ToggleInput = shallowUntil(
      <Toggle
        input={{
          name: 'Awesome toggle',
          value: true,
          onChange: jest.fn(),
        }}
      />,
      {
        context: { intl },
        childContextTypes: { intl: intlShape },
      },
      'Toggle',
    );

    expect(ToggleInput).toMatchSnapshot();
  });

  it('renders a toggle with an error', () => {
    const ToggleInput = shallowUntil(
      <Toggle
        input={{
          name: 'Not very awesome toggle',
          value: false,
          onChange: jest.fn(),
        }}
        meta={{
          error: 'Ooops',
          touched: true,
        }}
      />,
      {
        context: { intl },
        childContextTypes: { intl: intlShape },
      },
      'Toggle',
    );

    expect(ToggleInput).toMatchSnapshot();
  });
});
