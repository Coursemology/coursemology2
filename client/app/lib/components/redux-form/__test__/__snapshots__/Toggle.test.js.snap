// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Toggle /> renders a toggle with an error 1`] = `
<div
  className="MuiFormControl-root MuiFormControl-fullWidth"
>
  <WithStyles(ForwardRef(FormControlLabel))
    control={
      <WithStyles(ForwardRef(Switch))
        checked={false}
        color="primary"
        onChange={[MockFunction]}
      />
    }
    label={<b />}
    name="Not very awesome toggle"
    style={
      Object {
        "height": "30px",
        "margin": "8px 0px 0px -8px",
      }
    }
  />
  <WithStyles(ForwardRef(FormHelperText))
    error={true}
    style={
      Object {
        "margin": 0,
      }
    }
  >
    Ooops
  </WithStyles(ForwardRef(FormHelperText))>
</div>
`;

exports[`<Toggle /> renders a toggled toggle 1`] = `
<div
  className="MuiFormControl-root MuiFormControl-fullWidth"
>
  <WithStyles(ForwardRef(FormControlLabel))
    control={
      <WithStyles(ForwardRef(Switch))
        checked={true}
        color="primary"
        onChange={[MockFunction]}
      />
    }
    label={<b />}
    name="Awesome toggle"
    style={
      Object {
        "height": "30px",
        "margin": "8px 0px 0px -8px",
      }
    }
  />
</div>
`;
